numpy/ma/core.py

```diff
@@ -3873,77 +3875,86 @@ def _delegate_binop(self, other):
                 return True
         return False
 
-    def __eq__(self, other):
-        """
-        Check whether other equals self elementwise.
+    def _comparison(self, other, compare):
+        """Compare self with other using operator.eq or operator.ne.
+
+        When either of the elements is masked, the result is masked as well,
+        but it is still set, with self and other considered equal if both
+        elements are masked, and unequal if only one of them is masked.
 
+        For structured arrays, all elements are combined, with masked values
+        ignored. The result is masked if all elements were masked, with the
+        self and other considered equal if all elements of both were masked.
         """
         if self is masked:
             return masked
+
         omask = getattr(other, '_mask', nomask)
-        if omask is nomask:
-            check = self.filled(0).__eq__(other)
-            try:
-                check = check.view(type(self))
-                check._mask = self._mask
-            except AttributeError:
-                # Dang, we have a bool instead of an array: return the bool
-                return check
+        smask = self._mask
+        if smask is nomask:
+            mask = omask
         else:
-            odata = filled(other, 0)
-            check = self.filled(0).__eq__(odata).view(type(self))
-            if self._mask is nomask:
-                check._mask = omask
-            else:
-                mask = mask_or(self._mask, omask)
-                if mask.dtype.names:
-                    if mask.size > 1:
-                        axis = 1
-                    else:
-                        axis = None
-                    try:
-                        mask = mask.view((bool_, len(self.dtype))).all(axis)
-                    except ValueError:
-                        mask = np.all([[f[n].all() for n in mask.dtype.names]
-                                       for f in mask], axis=axis)
-                check._mask = mask
+            mask = mask_or(smask, omask)
+
+        odata = other.data if isinstance(other, MaskedArray) else other
+        if mask.dtype.names and mask is not nomask:
+            # For possibly masked structured arrays we need to be careful,
+            # since the standard structured array comparison will use all
+            # elements, masked or not.  To avoid this, we set all masked
+            # elements in self to other, so they'll count as equal.
+            # To prepare, we ensure we have the right shape.
+            sbroadcast = np.broadcast_to(self, np.broadcast(self, other).shape,
```

Comment:
@eric-wieser - as you'll see, I had to leave the broadcasting alone, since otherwise I had to make a copy. I agree with you that, ideally, `filled` would just do the right thing itself. Even better if the `filled` function could take a `mask` rather than just calling the method.  But I think that's best left for another PR!